[ModuleInterface] Qualify all types in module interfaces

harlanhaskins opened the pull request:
We should not be making assumptions about module lookup when we're
compiling a module interface, so instead print the types fully
qualified.

rdar://48445154

harlanhaskins committed:
[ModuleInterface] Qualify all types in module interfaces

We should not be making assumptions about module lookup when we're
compiling a module interface, so instead print the types fully
qualified.

rdar://48445154

harlanhaskins committed:
[ModuleInterface] Allow falling back to prefer TypeReprs

There are still cases (a module with a type that's the same name as the
module) where we cannot fully qualify all types. In those cases, allow
them to remain unqualified with a flag, `-Xfrontend
-preserve-types-as-written-in-module-interface`.

harlanhaskins commented:
@swift-ci please test

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2530,10 +2542,17 @@ void PrintAST::visitVarDecl(VarDecl *decl) {
-    else
-      printTypeLoc(tyLoc);
+    } else {
+      // HACK: When printing result types for vars with opaque result types,
```
What is this part about?

harlanhaskins replied:
When we turned off PreferTypeRepr, we went through the full printing behavior for opaque result types. Since OpaqueReturnTypePrinting applies to every instances of opaque return types, it also replaced the return types of functions, properties, and subscripts with their stable representation, which isn’t what we want for the declaration itself. So we override the opaque return type printing while printing vars, funcs, and subscripts.

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2805,10 +2824,14 @@ void PrintAST::visitFuncDecl(FuncDecl *decl) {
+        PrintOptions subOptions = Options;
+        subOptions.OpaqueReturnTypePrinting =
+            PrintOptions::OpaqueReturnTypePrintingMode::WithOpaqueKeyword;
+        printTypeLocWithOptions(ResultTyLoc, subOptions);
```
Missing the "HACK" comment here?

Also, what happens when we support optional-wrapped opaque types? Can we handle both cases?

harlanhaskins replied:
I suppose we could just SaveAndRestore this setting for both implicitly unwrapped optional opaque result types and regular types. Wouldn't hurt, and would Just Work™ when we get around to it

jrose-apple commented on include/swift/Frontend/ParseableInterfaceSupport.h:
```diff
@@ -27,6 +27,10 @@ class ModuleDecl;
 struct ParseableInterfaceOptions {
+  /// Should we prefer printing TypeReprs when writing out types in a module
+  /// interface, or should we fully-qualify them?
+  bool PreserveTypesAsWrittenInModuleInterface = false;
```
I misread this as "types [as written in the module interface]" instead of "types [as written] [in the module interface]". Since it's already in ParseableInterfaceOptions, I think you can just drop the last part.

jrose-apple commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -535,6 +535,11 @@ def build_module_from_parseable_interface :
   ModeOpt;
 
+def preserve_types_as_written_in_module_interface :
+  Flag<["-"], "preserve-types-as-written-in-module-interface">,
```
…oh. Uh, "module-interface-preserve-types-as-written", maybe?

jrose-apple replied:
Can you add this to the XCTest overlay build, too, since we know we need it there?

jrose-apple commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -535,6 +535,11 @@ def build_module_from_parseable_interface :
+def preserve_types_as_written_in_module_interface :
+  Flag<["-"], "preserve-types-as-written-in-module-interface">,
+  HelpText<"When emitting a module interface, preserve the types as they were "
+           "written in the source, rather than fully-qualifying them.">;
```
No periods in help text. Also, a bit long for command-line help…

harlanhaskins committed:
Rename flag to ‘module-interface-preserve-types-as-written’

harlanhaskins committed:
Add new flag to XCTest overlay

harlanhaskins commented:
@swift-ci please test

harlanhaskins committed:
[Sema] Stop filtering nested protocol types from lookup

This check wasn't ever correct, because the fact that the the protocol comes
from another module doesn't change the fact that the type is valid for lookup
within this module. It incorrectly rejects the following, valid code:

```swift
// In A.swift
public protocol A {}
```

```
// In B.swift
import A

extension A {
  typealias B = Int
  func b(_ b: Self.B) {}
}
```

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test source compatibility

harlanhaskins commented:
`ERROR: Workspace has a .git repository, but it appears to be corrupt.`

@swift-ci please test source compatibility

harlanhaskins commented:
@swift-ci please test source compatibility
